### ISSUE TITLE
feat(torch): add a GPU FDTD backend while preserving the C++ backend

### ISSUE THREAD

**Issue by ruddyscent** [ruddyscent/gmes#75]
## 배경

GMES의 시간 진행 핵심은 NumPy field를 SWIG typemap으로 C++ `update_all()`에 전달하는 구조다. 현재 C++ 경로는 다음 책임이 서로 결합되어 있다.

- `gmes/fdtd.py`: field 할당, piecewise material/source mapping, component update scheduling, Bloch/MPI halo 교환
- `src/material.pyx`: material 계수 계산과 C++ pointwise updater/parameter 생성
- `src/pw_*.hh`, `src/pw_material.i`: 재료별 C++ update kernel과 NumPy/SWIG 경계
- `gmes/pw_source.py`, `gmes/source.py`: point/TFSF/Gaussian source와 auxiliary FDTD
- `gmes/file_io.py`, `gmes/show.py`: NumPy 기반 출력과 시각화

GPU 가속을 위해 기존 C++ 백엔드를 제거하지 않고 PyTorch 백엔드를 병렬로 추가한다. C++는 계속 기본 실행 경로이자 수치 동등성의 기준 구현으로 사용한다. NumPy solver backend는 별도로 만들지 않는다.

PyTorch의 공식 설치 범위에는 Python 3.14가 포함되므로 현재 프로젝트의 Python 요구사항과 양립할 수 있다. 실제 최소 PyTorch 버전과 CPU/CUDA wheel 조합은 CI probe 후 고정한다.

- https://pytorch.org/get-started/locally/
- https://github.com/pytorch/pytorch/blob/main/RELEASE.md

## 목표

- 기존 생성자와 `backend="cpp"` 동작을 변경하지 않는다.
- 단일 프로세스에서 PyTorch CPU와 CUDA backend를 명시적으로 선택할 수 있게 한다.
- 현재 piecewise material 의미와 공개 material/geometry/source API를 보존한다.
- field, 계수, material state가 time-step 동안 선택한 device에 상주하게 한다.
- C++ ↔ Torch CPU ↔ CUDA 수치 동등성을 비영 입력과 다중 step 상태까지 검증한다.
- 재현 가능한 성능·메모리 benchmark로 GPU 이점과 crossover 크기를 문서화한다.

## MVP 비목표

- C++/SWIG/Cython backend 제거
- NumPy time-stepping backend 추가
- autograd 또는 differentiable FDTD
- 다중 GPU, CUDA-aware MPI 또는 NCCL halo exchange
- 첫 단계부터 `torch.compile`, Triton, custom C++/CUDA op 적용
- 미지원 material/source를 C++로 조용히 fallback

미지원 조합은 `init()` 중 명확한 오류를 내야 한다. 특히 MVP에서 `backend="torch"`와 `space.numprocs > 1` 조합은 거부한다.

## 제안 API

기존 호출은 C++를 계속 사용한다.

```python
simulation = TMzFDTD(space, geometry, sources)
```

Torch는 명시적으로 선택한다.

```python
simulation = TMzFDTD(
    space,
    geometry,
    sources,
    backend="torch",
    device="cuda",
    dtype="float64",
    compile=False,
)
```

계약:

- 기본 backend는 `cpp`다.
- `torch`는 optional extra와 lazy import로 제공한다.
- Torch가 없거나 요청한 device가 없으면 자동 fallback하지 않는다.
- 동등성 기본 dtype은 real `float64`, Bloch/complex `complex128`이다.
- `float32`/`complex64`는 검증 후 명시적 성능 모드로 추가한다.
- `float16`/`bfloat16`은 초기 지원 범위에서 제외한다.
- Torch field는 backend-native `torch.Tensor`로 노출한다.
- host 변환은 `field_numpy(component, copy=True)` 같은 명시적 API로만 제공한다.

### 1. Backend composition

`FDTD`의 스케줄은 유지하고 다음 최소 계약을 갖는 backend 객체를 주입한다.

```text
FDTD
 ├── CppBackend       # 현재 NumPy + SWIG/C++ 경로
 └── TorchBackend
      ├── FieldStorage
      ├── MaterialPlan
      ├── SourcePlan
      ├── BoundaryExchange
      └── HostAdapter
```

Backend 책임:

- field 할당과 dtype/device 검증
- material/source plan 생성과 finalize
- E/H component update
- serial periodic/Bloch boundary 처리
- clone/deepcopy와 serialization 정책
- host array 변환과 synchronization

C++ backend는 기존 `get_pw_material_*()`, `merge()`, `update_all()` 경로를 가능한 한 그대로 감싼다. backend 도입 자체가 native hot path를 느리게 해서는 안 된다.

### 2. Backend-neutral piecewise lowering

현재 `src/material.pyx`가 계수 계산과 C++ updater 생성을 함께 수행하는 결합을 분리한다.

```text
geometry lookup
  → backend-neutral point record
  → material/component/state-shape별 CPU bucket
  → target uniqueness/coverage/bounds validation
  → finalize(device, dtype)
  → backend execution plan
```

- geometry mapping 중에는 Python list 또는 NumPy host storage에만 수집한다.
- 점마다 Torch tensor를 만들거나 GPU로 전송하지 않는다.
- material에 Torch 코드를 직접 넣지 않고 lowering registry/factory를 둔다.
- 기존 PML 및 분산 재료 계수 계산식을 공유하고 backend별로 재유도하지 않는다.
- 사용자 정의 material은 Torch lowerer 등록 지점을 제공하고, 미등록 material은 fail-fast한다.

### 3. Tensor layout과 dense/sparse 경로

- 기존 Yee layout을 유지해 `ex`, `ey`, `ez`, `hx`, `hy`, `hz`를 서로 다른 shape의 contiguous 3-D tensor로 둔다.
- 비활성 component의 `(1, 1, 1)` storage와 1-D/2-D collapsed-axis 의미를 보존한다.
- canonical host index는 `(N, 3)`, GPU 실행 인덱스는 memory 절약을 위해 flattened target/source offset을 사용한다.
- coefficient는 structure-of-arrays, persistent state는 field dtype tensor로 보관한다.
- target offset은 unique임을 초기화 시 검증한다. 중복 write는 backend별 의미가 달라질 수 있다.

첫 정책은 hybrid로 고정한다.

- Dense: DefaultMedium을 포함한 비분산 Dielectric. component별 coefficient grid와 contiguous stencil을 사용한다.
- Sparse: Const, UPML, CPML, Drude, Lorentz, DCP, DM2 및 불규칙 영역. 미리 계산한 flat offset으로 gather하고 unique target에 write한다.
- pole/transition 수가 다른 재료는 ragged tensor 대신 state shape별 bucket으로 나눈다.

권장 bucket key:

- Drude/Lorentz: `(kind, component, pole_count)`
- DCP: `(kind, component, drude_count, critical_point_count)`
- DM2: `(kind, component, transition_count)`

### 4. 실행 정책

- 첫 구현은 eager + `torch.no_grad()` + in-place update로 정확성을 확보한다.
- `requires_grad=True` 입력은 명시적으로 거부한다.
- 정상 `step()` 경로에서는 `.cpu()`, `.numpy()`, `.item()` 및 NumPy↔Torch 변환을 금지한다.
- `torch.compile`은 eager parity 후 material kernel 단위로 평가한다. geometry, I/O, MPI는 compiled region에 넣지 않는다.
- profiler로 graph break, kernel count, host/device transfer를 확인한 뒤에만 Triton/custom op을 검토한다.

### 5. Source, I/O와 MPI

- PointSource/current source는 component별 target/value batch로 lowering한다.
- TFSF/Transparent source는 겹치는 face 기여를 target별로 합산한 뒤 unique write한다.
- auxiliary FDTD는 main solver의 backend/device/dtype을 반드시 상속한다.
- probe는 device buffer에 모아 일괄 전송해 매 step GPU sync를 피한다.
- `write_field`, snapshot, plotting은 명시적 `detach().cpu().numpy()` 경계에서 처리하고 solver benchmark에서 제외한다.
- 다중 GPU는 별도 후속 이슈로 분리한다. 장기적으로는 1 process/1 GPU, NCCL P2P halo exchange를 검토하며 DDP는 사용하지 않는다.

## 구현 단계

각 단계는 별도 PR로 제출한다.

### Phase 0: 계약과 기준선

- [ ] backend/device/dtype 공개 API와 지원 matrix를 확정한다.
- [ ] 현재 C++ `float64`/`complex128` 결과를 golden oracle로 고정한다.
- [ ] seeded nonzero field를 사용하는 backend parity helper를 추가한다.
- [ ] 초기화, steady-state step, memory를 분리 측정하는 benchmark harness를 추가한다.
- [ ] unsupported feature의 fail-fast 정책을 정의한다.

### Phase 1: Backend 기반

- [ ] `CppBackend`와 `TorchBackend` 최소 계약을 구현한다.
- [ ] optional Torch dependency와 lazy import/error 안내를 추가한다.
- [ ] field allocation, host conversion, clone/deepcopy를 backend dispatch한다.
- [ ] Yee component별 stencil/collapsed-axis metadata를 정의한다.
- [ ] backend-neutral piecewise record builder와 one-shot device finalize를 구현한다.
- [ ] index bounds, unique target, 전체 material coverage를 검증한다.

### Phase 2: Torch MVP

- [ ] Dummy, Const, Dielectric의 6개 component를 구현한다.
- [ ] TE/TM/TEM 및 1-D/2-D/3-D를 지원한다.
- [ ] PointSource와 electric/magnetic current source를 구현한다.
- [ ] real `float64`, Bloch `complex128`, serial periodic boundary를 구현한다.
- [ ] Torch CPU와 CUDA 결과를 C++와 비교한다.

### Phase 3: 흡수 경계

- [ ] UPML state를 구현하고 반복 갱신을 검증한다.
- [ ] CPML `psi1`/`psi2` state를 구현하고 반복 갱신을 검증한다.
- [ ] state clone/serialization을 지원한다.
- [ ] 장시간 reflection regression을 추가한다.

### Phase 4: 분산 재료

- [ ] Drude를 구현한다.
- [ ] Lorentz를 구현한다.
- [ ] DcpAde를 구현한다.
- [ ] DcpPlrc와 DcpRc를 구현한다.
- [ ] real-only DM2를 마지막에 구현한다.
- [ ] pole/transition 수가 다른 piecewise bucket을 검증한다.

DM2는 host sync 없이 active mask로 cell별 predictor-corrector 수렴 상태를 관리하고, 현재 iteration limit와 zero-reference error 의미를 보존해야 한다.

### Phase 5: 고급 source와 사용자 기능

- [ ] TFSF/Transparent source를 구현한다.
- [ ] GaussianBeam과 auxiliary FDTD backend 전파를 구현한다.
- [ ] probe, `write_field`, snapshot, plotting의 host adapter를 구현한다.
- [ ] public field access, deepcopy, pickle/checkpoint 동작을 문서화하고 검증한다.

### Phase 6: 최적화와 배포

- [ ] eager kernel을 기준으로 `torch.compile`을 선택적으로 도입한다.
- [ ] `float32`/`complex64` 성능 모드를 검증한다.
- [ ] dense/sparse threshold와 material bucket launch overhead를 측정한다.
- [ ] GPU memory와 flat offset storage를 최적화한다.
- [ ] README, 지원 matrix, GPU example, benchmark 재현 문서를 추가한다.

## 수치 검증

현재 `test_pw_*`의 0-field 중심 테스트만으로는 충분하지 않다. 동일한 seeded nonzero field와 동일한 material mapping을 C++와 Torch에 공급해 다음을 비교한다.

- 6개 Yee component
- 1-D/2-D/3-D와 singleton axis
- real/complex
- 단일 점, 여러 점, `merge()`된 서로 다른 계수
- 1/2/5/20/100 step
- UPML/CPML/pole/critical-point/DM2 persistent state
- underlying material의 `eps_inf`/`mu_inf`
- 중첩 source의 순차 적용 의미

초기 허용 오차 제안:

| 비교 | 기준 |
|---|---|
| shape, dtype, device, index 수/소속 | exact |
| Torch CPU double, 단일 kernel | `rtol=1e-12`, `atol=1e-13` |
| CUDA double, 단일 kernel | `rtol=1e-11`, `atol=1e-12` |
| double, 최대 100-step | 상대 L2 `<=1e-9`, scaled L-infinity `<=1e-8` |
| float32/complex64, 단일 kernel | `rtol=2e-5`, `atol=2e-6` |
| CPML reflection energy | C++ 대비 상대 차이 `<=1%` |

허용 오차는 구현 초기에 모델별로 측정하고, 완화가 필요하면 근거와 함께 고정한다. 장시간 검증은 pointwise 값뿐 아니라 probe spectrum, 반사율, 투과율과 에너지를 함께 사용한다.

## Benchmark 전략

Microbenchmark:

- Dense Dielectric: `64^3`, `128^3`, `256^3`
- sparse material coverage: 1%, 10%, 50%, 100%
- CPML shell
- Drude/Lorentz 1 pole와 4 poles
- DCP와 DM2
- eager와 compile을 별도 측정

End-to-end:

- `air2d`: 작은 문제의 GPU crossover
- 축소/전체 3-D mixed-material 사례
- `metal_array --quick`: DCP + CPML
- 축소 Fresnel/Bloch: complex + dispersive material
- `tfsf_with_scatterer`: auxiliary solver와 transparent source

측정 규칙:

- mapping, tensor packing, H2D, compile cold start, steady-state를 분리한다.
- warm-up 후 CUDA synchronize 경계를 명시하고 median/p95를 기록한다.
- steps/s, updated cells/s, CPU RSS, CUDA allocated/reserved peak를 기록한다.
- hardware, OS, PyTorch/CUDA, dtype, compile 상태를 함께 기록한다.
- 파일 I/O와 시각화는 solver timing에서 제외한다.

성능 목표:

- backend abstraction으로 인한 기존 C++ 경로 성능 저하는 5% 이내다.
- Torch step 중 암묵적 host/device transfer가 없다.
- warm-up 이후 device memory가 step 수에 따라 증가하지 않는다.
- 문서화된 대형 3-D 사례에서 CUDA가 C++ single-process보다 최소 2배 빠른 것을 목표로 하되, GPU별 FP64 성능 차이 때문에 보편적인 hard gate로 사용하지 않는다.
- 작은 문제에서 GPU가 느려지는 crossover를 숨기지 않고 문서화한다.

## CI와 패키징

- [ ] 기존 required job 이름과 C++ 테스트를 유지한다.
- [ ] Torch 없는 base 설치에서 C++ import/smoke/build를 검증한다.
- [ ] Torch CPU parity 소형 테스트를 별도 CI job에서 실행한다.
- [ ] NVIDIA GPU runner에서 nightly 또는 `workflow_dispatch` CUDA 검증을 추가한다.
- [ ] CUDA runner에서 `torch.cuda.is_available()`가 false면 skip하지 않고 실패시킨다.
- [ ] wheel/sdist에서 Torch는 optional dependency로 유지한다.
- [ ] 지원하는 Python/PyTorch/accelerator wheel matrix를 문서화한다.

## 완료 조건

- [ ] 기존 호출과 `backend="cpp"`에서 API·수치 결과·전체 테스트가 유지된다.
- [ ] `backend="torch"`가 CPU와 CUDA에서 명시적으로 선택 가능하다.
- [ ] Torch CPU double이 C++와 kernel 및 end-to-end 허용 오차 내 일치한다.
- [ ] CUDA double이 Torch CPU와 허용 오차 내 일치한다.
- [ ] real/complex, 1-D/2-D/3-D, 전 component/mode가 검증된다.
- [ ] PML과 분산 material state의 다중 step parity가 검증된다.
- [ ] 정상 `step()`에 host/device 변환과 step별 memory 증가가 없다.
- [ ] 미지원 material/source/MPI 조합이 초기화 시 명확히 실패한다.
- [ ] 재현 가능한 2-D/3-D/mixed/dispersive benchmark 결과가 게시된다.
- [ ] 설치, backend 선택, dtype trade-off, 지원 범위와 제한이 문서화된다.
